eclipse-theia/theia
# Revise Git watchers

#### What it does

Current Git watchers have several problems:
 - `sync` calls for not-watching watchers leave them in watching state
 - `git status` runs may supersede each other
 - singleton repository manager depending on connection scoped git

Fixes #4436

#### How to test
- there's some debug logging added in a separate commit to this PR, it helps to see what's going on behind the scene.
- test cases:
  1. open a workspace with two git repositories
      - [ ] see only one watcher is running; try switch to second repo
  2. open a second frontend and make sure to have the same git repo selected as in first one
      - [ ] see only one watcher is running
      - [ ] switch to the other git repo in both; see only one watcher is running
      - [ ] have different repos selected; see two watchers are running


#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
